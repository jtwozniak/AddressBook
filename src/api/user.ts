const UserPerRequest = 50
const MaxUsers = 1000 // users limit, api support up to 5000
const LastPage = MaxUsers / UserPerRequest

export const fetchUsers = async (
  nat: string,
  page: number
): Promise<User[]> => {
  // const paramsObject =

  if (page > LastPage) {
    return new Promise((resolve, reject) => {
      resolve([] as User[])
    })
  }

  const params = new URLSearchParams({
    page: String(page),
    results: String(UserPerRequest),
    inc: ["picture", "name", "email", "location", "phone", "cell"].join(","), // download only required fields
    seed: "Sherpany",
  })

  if (nat) {
    params.append("nat", nat)
  }

  console.log("Api call", nat, page)

  // This could by typed better, but there is not point to do that here
  // usually there would be a way of generating types from API - like from swagger or graphql
  const data: any = await fetch(
    `https://randomuser.me/api/?${params.toString()}`
  ).then((data) => data.json())
  const users = data.results.map(
    ({
      picture: { thumbnail: picture },
      name,
      email,
      phone,
      cell,
      location: { street, city, state, postcode },
    }: any) => ({
      picture,
      name,
      email,
      phone,
      cell,
      location: { street, city, state, postcode },
    })
  )

  return users
}

export type User = {
  picture: string // thumbnail
  name: { title: string; first: string; last: string } // Code test have error, the field username doesn't exist - I assume title instead
  email: string
  location: {
    street: {
      name: string
      number: number
    }
    city: string
    state: string
    postcode: number
  }
  phone: string
  cell: string
}
