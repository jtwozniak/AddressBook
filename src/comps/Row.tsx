import React from "react"
import { User } from "../api/user"

export const Row = (props: User) => {
  return (
    <tr key={props.cell}>
      <td>{JSON.stringify(props)}</td>
    </tr>
  )
}
