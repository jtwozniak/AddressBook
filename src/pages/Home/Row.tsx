import React, { useState } from "react"
import styled from "styled-components"
import { User } from "../../api/user"

const Tr = styled.tr<{
  // $ sing in styled-components means to not pass prop to dom
  // https://styled-components.com/docs/api#transient-props
  $mouseOver: boolean
}>`
  cursor: pointer;
  opacity: ${({ $mouseOver }) => ($mouseOver ? 1 : 0.6)};
`

type Props = User & {
  setModalUser: (u: User) => void
}

export const Row = ({ setModalUser, ...user }: Props) => {
  const [mouseOver, setMouseOver] = useState(false)

  return (
    <Tr
      data-testid="row"
      key={user.cell}
      $mouseOver={mouseOver}
      onMouseOver={() => {
        if (!mouseOver) setMouseOver(true)
      }}
      onMouseLeave={() => {
        if (mouseOver) setMouseOver(false)
      }}
      onClick={() => {
        setModalUser(user)
      }}
    >
      <td>
        <img src={user.picture} />
      </td>
      <td>{user.name.title}</td>
      <td>{user.name.first}</td>
      <td>{user.name.last}</td>
      <td>{user.email}</td>
    </Tr>
  )
}
