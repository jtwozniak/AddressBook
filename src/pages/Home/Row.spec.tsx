import React from "react"
import { act, fireEvent, render } from "@testing-library/react"
import { User } from "../../api/user"
import { Row } from "./Row"
// for some reason setupFilesAfterEnv didn't import it automatically
// I could create separate file for that but that is not worth it, at least not for one test
import "jest-styled-components"

const user: User = {
  picture: "thumbnail",
  name: { title: "title", first: "first", last: "last" },
  email: "string",
  location: {
    street: {
      name: "name",
      number: 1,
    },
    city: "city",
    state: "state",
    postcode: 90210,
  },
  phone: "phone",
  cell: "cell",
}

describe("Row", () => {
  it("hover style test", () => {
    const { container, getByTestId } = render(
      <Row setModalUser={jest.fn()} {...user} />,
      {
        wrapper: ({ children }) => (
          <table>
            <tbody>{children}</tbody>
          </table>
        ),
      }
    )

    expect(container.firstChild).toMatchSnapshot()

    const rowComponent = getByTestId("row")
    // toHaveStyleRule comes from jest-styled-components
    expect(rowComponent).toHaveStyleRule("opacity", "0.6")

    act(() => {
      fireEvent.mouseOver(rowComponent)
    })

    expect(rowComponent).toHaveStyleRule("opacity", "1")
  })
})
