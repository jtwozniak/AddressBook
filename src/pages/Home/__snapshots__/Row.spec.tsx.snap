// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row hover style test 1`] = `
.c0 {
  cursor: pointer;
  opacity: 0.6;
}

<table>
  <tbody>
    <tr
      class="c0"
      data-testid="row"
    >
      <td>
        <img
          src="thumbnail"
        />
      </td>
      <td>
        title
      </td>
      <td>
        first
      </td>
      <td>
        last
      </td>
      <td>
        string
      </td>
    </tr>
  </tbody>
</table>
`;
